File: api/infer-lineage.js
/**
 * api/infer-lineage.js — Vercel serverless function.
 *
 * Takes the source queries a person has pasted/uploaded and asks the
 * configured LLM (see lib/llmClient.js / .env.example — OpenRouter by
 * default, or a direct/custom/local model) to extract table/column-
 * level lineage from them. Runs only on Vercel's servers — no API key
 * of any kind ever reaches the browser.
 */

'use strict';

const { callLLM, parseJsonResponse, LLMConfigError, LLMUpstreamError } = require('../lib/llmClient');
const { buildLineagePrompt } = require('../lib/prompts');
const { validateAndNormalizeGraph, GraphValidationError } = require('../lib/validateGraph');
const { checkRateLimit, clientKeyFromRequest, ensureSweepScheduled } = require('../lib/rateLimit');
const { isOriginAllowed } = require('../lib/checkOrigin');

const MAX_SOURCES = 20;
const MAX_TOTAL_CONTENT_CHARS = 80_000; // generous for real SSAS query sets, cheap to enforce
const MAX_BODY_BYTES = 400 * 1024; // pasted queries can legitimately be large; still far below Vercel's ~4.5MB platform limit
const ALLOWED_DIALECTS = new Set(['sql', 'dax', 'm', 'other']);

// This endpoint costs a real LLM call per request, so it gets a
// tighter budget than a read-only endpoint would. See lib/rateLimit.js
// for what this protection does and doesn't cover.
const RATE_LIMIT = { windowMs: 60 * 1000, max: 6 };
const GLOBAL_RATE_LIMIT_KEY = '__global_infer__';
const GLOBAL_RATE_LIMIT = {
  windowMs: 60 * 1000,
  max: Number(process.env.GLOBAL_RATE_LIMIT_PER_MINUTE) || 60,
};

/**
 * @param {import('http').IncomingMessage} req
 * @param {import('http').ServerResponse} res
 */
module.exports = async function handler(req, res) {
  if (req.method !== 'POST') {
    res.status(405).json({ error: 'Method not allowed' });
    return;
  }

  const contentLength = Number(req.headers['content-length'] || 0);
  if (contentLength > MAX_BODY_BYTES) {
    res.status(413).json({ error: 'Request body is too large.' });
    return;
  }
  if (!isOriginAllowed(req)) {
    res.status(403).json({ error: 'Origin not allowed.' });
    return;
  }

  ensureSweepScheduled(Math.max(RATE_LIMIT.windowMs, GLOBAL_RATE_LIMIT.windowMs));
  const clientKey = clientKeyFromRequest(req);
  const rate = checkRateLimit(clientKey, RATE_LIMIT);
  res.setHeader('X-RateLimit-Limit', String(RATE_LIMIT.max));
  res.setHeader('X-RateLimit-Remaining', String(rate.remaining));
  if (!rate.allowed) {
    res.setHeader('Retry-After', String(rate.retryAfterSeconds));
    res.status(429).json({ error: 'Too many requests — please slow down and try again shortly.' });
    return;
  }
  const globalRate = checkRateLimit(GLOBAL_RATE_LIMIT_KEY, GLOBAL_RATE_LIMIT);
  if (!globalRate.allowed) {
    res.setHeader('Retry-After', String(globalRate.retryAfterSeconds));
    res.status(429).json({ error: 'This deployment is getting more traffic than usual — please try again shortly.' });
    return;
  }

  const sources = Array.isArray(req.body?.sources) ? req.body.sources : null;
  if (!sources || sources.length === 0) {
    res.status(400).json({ error: 'Provide at least one source (sources: [{ name, dialect, content }]).' });
    return;
  }
  if (sources.length > MAX_SOURCES) {
    res.status(400).json({ error: `Too many sources — the limit is ${MAX_SOURCES} per request.` });
    return;
  }

  let totalChars = 0;
  const cleanSources = [];
  for (const s of sources) {
    if (!s || typeof s.content !== 'string' || !s.content.trim()) {
      res.status(400).json({ error: 'Each source needs non-empty "content".' });
      return;
    }
    totalChars += s.content.length;
    cleanSources.push({
      name: typeof s.name === 'string' && s.name.trim() ? s.name.trim().slice(0, 200) : 'untitled',
      dialect: ALLOWED_DIALECTS.has(s.dialect) ? s.dialect : 'other',
      content: s.content,
    });
  }
  if (totalChars > MAX_TOTAL_CONTENT_CHARS) {
    res.status(400).json({ error: `Combined source content is too large (limit ${MAX_TOTAL_CONTENT_CHARS} characters).` });
    return;
  }

  try {
    const { system, userContent } = buildLineagePrompt(cleanSources);
    const responseText = await callLLM({
      system,
      messages: [{ role: 'user', content: userContent }],
      maxTokens: 8000,
      temperature: 0,
    });

    let rawGraph;
    try {
      rawGraph = parseJsonResponse(responseText);
    } catch {
      res.status(502).json({ error: 'The model did not return valid JSON for this source set. Try again, or split large sources into smaller pieces.' });
      return;
    }

    const { graph, warnings } = validateAndNormalizeGraph(rawGraph, cleanSources.map((s) => s.name));
    res.status(200).json({ graph, warnings });
  } catch (err) {
    if (err instanceof LLMConfigError) {
      res.status(500).json({ error: err.message });
      return;
    }
    if (err instanceof LLMUpstreamError) {
      res.status(err.status).json({ error: err.message });
      return;
    }
    if (err instanceof GraphValidationError) {
      res.status(502).json({ error: `The model's output didn't match the expected graph shape: ${err.message}` });
      return;
    }
    console.error('Unexpected error in /api/infer-lineage:', err);
    res.status(500).json({ error: 'Unexpected server error.' });
  }
};

File: api/infer-lineage.py
"""
api/infer-lineage.py — Vercel Python serverless function.

The core "AI" for lineage extraction, replaced end to end: this parses
pasted SQL/DAX/M with a real SQL parser (sqlglot) plus documented
heuristics for DAX and M (see api/_lineage_engine.py) and returns a
lineage graph. No LLM call, no API key, nothing leaves this function.
/api/ask and /api/report are unrelated and still optionally use an LLM
(see lib/llmClient.js) — this endpoint never did and no longer can.
"""

import json
import os
import time
from http.server import BaseHTTPRequestHandler

from _lineage_engine import extract_lineage

MAX_SOURCES = 20
MAX_TOTAL_CONTENT_CHARS = 80_000
MAX_BODY_BYTES = 400 * 1024
ALLOWED_DIALECTS = {'sql', 'dax', 'm', 'other'}

# Local parsing has no per-request $ cost to protect (unlike the old
# LLM-backed version), so this is a much looser, DoS-hygiene-only
# limit rather than a budget guard — see lib/rateLimit.js for the
# fuller discussion of what an in-memory, single-instance limiter
# does and doesn't cover; the same caveats apply here.
RATE_LIMIT_WINDOW_SECONDS = 60
RATE_LIMIT_MAX = 30
_rate_buckets = {}  # {client_key: (window_start_epoch, count)}


def _client_key(headers):
    forwarded = headers.get('x-forwarded-for', '')
    if forwarded:
        return forwarded.split(',')[0].strip()
    return 'unknown'


def _check_rate_limit(key):
    now = time.time()
    window_start, count = _rate_buckets.get(key, (now, 0))
    if now - window_start >= RATE_LIMIT_WINDOW_SECONDS:
        _rate_buckets[key] = (now, 1)
        return True
    if count >= RATE_LIMIT_MAX:
        return False
    _rate_buckets[key] = (window_start, count + 1)
    return True


def _origin_allowed(headers):
    allowed = os.environ.get('ALLOWED_ORIGIN')
    if not allowed:
        return True
    origin = headers.get('origin')
    if not origin:
        return True
    return origin == allowed


class handler(BaseHTTPRequestHandler):
    def _send_json(self, status, payload):
        body = json.dumps(payload).encode('utf-8')
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_POST(self):
        if not _origin_allowed(self.headers):
            self._send_json(403, {'error': 'Origin not allowed.'})
            return

        client_key = _client_key(self.headers)
        if not _check_rate_limit(client_key):
            self._send_json(429, {'error': 'Too many requests — please slow down and try again shortly.'})
            return

        try:
            declared_length = int(self.headers.get('content-length', 0))
        except ValueError:
            declared_length = 0
        if declared_length > MAX_BODY_BYTES:
            self._send_json(413, {'error': 'Request body is too large.'})
            return

        # Read at most MAX_BODY_BYTES + 1 regardless of the declared
        # header, so a missing or dishonest Content-Length can't be
        # used to stream an unbounded body past the check above.
        raw = self.rfile.read(min(declared_length, MAX_BODY_BYTES + 1) or MAX_BODY_BYTES + 1)
        if len(raw) > MAX_BODY_BYTES:
            self._send_json(413, {'error': 'Request body is too large.'})
            return

        try:
            payload = json.loads(raw.decode('utf-8'))
        except (json.JSONDecodeError, UnicodeDecodeError):
            self._send_json(400, {'error': 'Request body must be valid JSON.'})
            return

        sources = payload.get('sources') if isinstance(payload, dict) else None
        if not isinstance(sources, list) or len(sources) == 0:
            self._send_json(400, {'error': 'Provide at least one source (sources: [{ name, dialect, content }]).'})
            return
        if len(sources) > MAX_SOURCES:
            self._send_json(400, {'error': f'Too many sources — the limit is {MAX_SOURCES} per request.'})
            return

        clean_sources = []
        total_chars = 0
        for s in sources:
            if not isinstance(s, dict) or not isinstance(s.get('content'), str) or not s.get('content', '').strip():
                self._send_json(400, {'error': 'Each source needs non-empty "content".'})
                return
            content = s['content']
            total_chars += len(content)
            name = s.get('name')
            name = name.strip()[:200] if isinstance(name, str) and name.strip() else 'untitled'
            dialect = s.get('dialect') if s.get('dialect') in ALLOWED_DIALECTS else 'other'
            clean_sources.append({'name': name, 'dialect': dialect, 'content': content})

        if total_chars > MAX_TOTAL_CONTENT_CHARS:
            self._send_json(400, {'error': f'Combined source content is too large (limit {MAX_TOTAL_CONTENT_CHARS} characters).'})
            return

        try:
            graph, warnings = extract_lineage(clean_sources)
        except Exception as ex:  # a parsing bug should surface as a clean 500, not a raw traceback to the client
            self._send_json(500, {'error': f'Unexpected error while extracting lineage: {ex}'})
            return

        self._send_json(200, {'graph': graph, 'warnings': warnings})

    def do_GET(self):
        self._send_json(405, {'error': 'Method not allowed'})
